File: src/main/java/org/jukeboxmc/item/behavior/ItemFishingRod.java
package org.jukeboxmc.item.behavior;

import com.nukkitx.protocol.bedrock.data.SoundEvent;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.entity.Entity;
import org.jukeboxmc.entity.EntityType;
import org.jukeboxmc.entity.projectile.EntityFishingHook;
import org.jukeboxmc.event.entity.ProjectileLaunchEvent;
import org.jukeboxmc.item.Burnable;
import org.jukeboxmc.item.Durability;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;

import java.time.Duration;
import java.util.Objects;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class ItemFishingRod extends Item implements Durability, Burnable {

    public ItemFishingRod( Identifier identifier ) {
        super( identifier );
    }

    public ItemFishingRod( ItemType itemType ) {
        super( itemType );
    }

    @Override
    public boolean useInAir(@NotNull Player player, Vector clickVector ) {
        if ( player.getEntityFishingHook() == null ) {
            EntityFishingHook entityFishingHook = Objects.requireNonNull(Entity.create( EntityType.FISHING_HOOK ));
            entityFishingHook.setShooter( player );
            entityFishingHook.setLocation( player.getLocation().add( 0, player.getEyeHeight(), 0 ) );

            float force = 1.6f;
            entityFishingHook.setVelocity( new Vector(
                    (float) ( -Math.sin( Math.toRadians( player.getYaw() ) ) * Math.cos( Math.toRadians( player.getPitch() ) ) * force * force ),
                    (float) ( -Math.sin( Math.toRadians( player.getPitch() ) ) * force * force + 0.4f ),
                    (float) ( Math.cos( Math.toRadians( player.getYaw() ) ) * Math.cos( Math.toRadians( player.getPitch() ) ) * force * force ) ), false );

            ProjectileLaunchEvent projectileLaunchEvent = new ProjectileLaunchEvent( entityFishingHook, ProjectileLaunchEvent.Cause.SNOWBALL );
            player.getServer().getPluginManager().callEvent( projectileLaunchEvent );
            if ( !projectileLaunchEvent.isCancelled() ) {
                entityFishingHook.spawn();
                this.updateItem( player, 1 );
                player.setEntityFishingHook( entityFishingHook );
                player.getWorld().playSound( player.getLocation(), SoundEvent.THROW, -1, "minecraft:player", false, false );
                return true;
            }
            return false;
        } else {
            player.getEntityFishingHook().close();
            player.setEntityFishingHook( null );
        }
        return true;
    }

    @Override
    public Duration getBurnTime() {
        return Duration.ofMillis( 300 );
    }

    @Override
    public int getMaxDurability() {
        return 384;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/item/behavior/ItemFishingRod.kt
package org.jukeboxmc.server.item.behavior

import org.cloudburstmc.protocol.bedrock.data.SoundEvent
import org.jukeboxmc.api.event.entity.ProjectileLaunchEvent
import org.jukeboxmc.api.item.Durability
import org.jukeboxmc.api.item.ItemType
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.entity.projectile.JukeboxEntityFishingHook
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import kotlin.math.cos
import kotlin.math.sin

class ItemFishingRod(itemType: ItemType, countNetworkId: Boolean) : JukeboxItem(itemType, countNetworkId), Durability {

    override fun useInAir(player: JukeboxPlayer, clickVector: Vector): Boolean {
        if (player.getEntityFishingHook() == null) {
            val force = 1.1f
            val entityFishingHook = JukeboxEntityFishingHook()
            entityFishingHook.setLocation(player.getLocation().add(0.5f, player.getEyeHeight(), 0.5f))
            entityFishingHook.setMotion(Vector(
                (-sin(Math.toRadians(player.getYaw().toDouble())) * cos(Math.toRadians(player.getPitch().toDouble())) * force * force).toFloat(),
                    (-sin(Math.toRadians(player.getPitch().toDouble())) * force * force + 0.4f).toFloat(),
                    (cos(Math.toRadians(player.getYaw().toDouble())) * cos(Math.toRadians(player.getPitch().toDouble())) * force * force).toFloat()
                ))
            entityFishingHook.setShooter(player)

            val projectileLaunchEvent = ProjectileLaunchEvent(entityFishingHook, ProjectileLaunchEvent.Cause.FISHING_ROD)
            JukeboxServer.getInstance().getPluginManager().callEvent(projectileLaunchEvent)
            if (projectileLaunchEvent.isCancelled()) return false
            entityFishingHook.spawn()
            this.updateDurability(player, 1)
            player.setEntityFishingHook(entityFishingHook)
            player.getWorld().playLevelSound(player.getLocation(), SoundEvent.THROW, -1, "minecraft:player")
            return true
        } else {
            player.getEntityFishingHook()?.remove()
            player.setEntityFishingHook(null)
        }
        return true
    }

    override fun getMaxDurability(): Int {
        return 384
    }

}